src/ImageSharp.Formats.Tiff/TiffDecoderCore.cs

```diff
@@ -0,0 +1,452 @@
+        /// <summary>
+        /// The decoder options.
+        /// </summary>
+        private readonly IDecoderOptions options;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TiffDecoderCore" /> class.
+        /// </summary>
+        /// <param name="options">The decoder options.</param>
+        public TiffDecoderCore(IDecoderOptions options)
+        {
+            this.options = options ?? new DecoderOptions();
+        }
+
+        public TiffDecoderCore(Stream stream, bool isLittleEndian, IDecoderOptions options) : this(options)
+        {
+            this.InputStream = stream;
+            this.IsLittleEndian = isLittleEndian;
+        }
+
+        /// <summary>
+        /// Gets the input stream.
+        /// </summary>
+        public Stream InputStream { get; private set; }
+
+        /// <summary>
+        /// A flag indicating if the file is encoded in little-endian or big-endian format.
+        /// </summary>
+        public bool IsLittleEndian;
+
+        /// <summary>
+        /// Decodes the image from the specified <see cref="Stream"/>  and sets
+        /// the data to image.
+        /// </summary>
+        /// <typeparam name="TColor">The pixel format.</typeparam>
+        /// <param name="image">The image, where the data should be set to.</param>
+        /// <param name="stream">The stream, where the image should be.</param>
+        /// <param name="metadataOnly">Whether to decode metadata only.</param>
+        public void Decode<TColor>(Image<TColor> image, Stream stream, bool metadataOnly)
+            where TColor : struct, IPixel<TColor>
+        {
+            this.InputStream = stream;
+
+            uint firstIfdOffset = ReadHeader();
+            TiffIfd firstIfd = ReadIfd(firstIfdOffset);
+        }
+
+        /// <summary>
+        /// Dispose
+        /// </summary>
+        public void Dispose()
+        {
+        }
+
+        public uint ReadHeader()
+        {
+            byte[] headerBytes = new byte[TiffConstants.SizeOfTiffHeader];
+            ReadBytes(headerBytes, TiffConstants.SizeOfTiffHeader);
+
+            if (headerBytes[0] == TiffConstants.ByteOrderLittleEndian && headerBytes[1] == TiffConstants.ByteOrderLittleEndian)
+                IsLittleEndian = true;
+            else if (headerBytes[0] != TiffConstants.ByteOrderBigEndian && headerBytes[1] != TiffConstants.ByteOrderBigEndian)
+                throw new ImageFormatException("Invalid TIFF file header.");
+
+            if (ToUInt16(headerBytes, 2) != TiffConstants.HeaderMagicNumber)
+                throw new ImageFormatException("Invalid TIFF file header.");
+
+            uint firstIfdOffset = ToUInt32(headerBytes, 4);
+            if (firstIfdOffset == 0)
+                throw new ImageFormatException("Invalid TIFF file header.");
+
+            return firstIfdOffset;
+        }
+
+        public TiffIfd ReadIfd(uint offset)
+        {
+            InputStream.Seek(offset, SeekOrigin.Begin);
```

Comment:
No we can't assume that. Someone could have multiple images in a stream or other information. Best we can do is look for the identifier at the position we are currently at.